File: tests/fixtures/go/kinds.go
// Package fixture holds one of every kind our Go tags.scm can emit.
//
// This is the fixture behind the kind-fidelity test in docs/plan.md M5:
// upstream's tags.scm emits @definition.type for all four of struct,
// interface, defined type and alias, and `type` is not in our Kind
// vocabulary at all, so every one of them would be an extractor guessing.
package fixture

// Config is a struct, with a documented field.
type Config struct {
	// Limit is how many entries to keep.
	Limit int
	quiet bool
}

// Mode is a defined type over a builtin. Distinct from an alias, which is the
// whole point: `type Mode int` introduces a new type and `type Key = string`
// does not.
type Mode int

// Key is a type alias.
type Key = string

// Handler is an interface, so a method with no body has somewhere to live.
type Handler interface {
	// Handle is the required method.
	Handle(key Key) bool
}

// Limit is a constant.
const Limit = 64

// Banner is a package-level variable.
var Banner = "codeintel"

// Handle satisfies Handler. Its receiver is the only thing that says so: the
// method is lexically at file scope and semantically owned by Config.
func (c Config) Handle(key Key) bool {
	return c.Limit > 0 && len(key) > 0
}

// describe is unexported, so `exported` must not hold for it.
func (c Config) describe() string {
	return Banner
}

File: tests/fixtures/rust/src/kinds.rs
//! Every kind our Rust `tags.scm` can emit, once each.
//!
//! This is the fixture behind the kind-fidelity test in `docs/plan.md` M2:
//! upstream's `tags.scm` maps `struct_item`, `enum_item`, `union_item` and
//! `type_item` all to `@definition.class`, so it fails this file.

/// A struct, with a documented field.
pub struct Config {
    /// How many entries to keep.
    pub limit: u32,
    quiet: bool,
}

/// An enum.
pub enum Mode {
    Fast,
    Careful,
}

/// A union. Distinct from an enum, which is the whole point.
pub union Word {
    bits: u32,
    halves: [u16; 2],
}

/// A type alias.
pub type Key = String;

/// A trait, with one required and one provided method.
pub trait Handler {
    /// Required.
    fn handle(&self, key: &Key) -> bool;

    /// Provided.
    fn name(&self) -> &'static str {
        "handler"
    }
}

/// A constant.
pub const LIMIT: u32 = 64;

/// A static.
pub static BANNER: &str = "codeintel";

impl Handler for Config {
    fn handle(&self, key: &Key) -> bool {
        self.limit > 0 && !key.is_empty()
    }
}

/// A macro.
#[macro_export]
macro_rules! shout {
    ($x:expr) => {
        format!("{}!", $x)
    };
}

/// A trait impl, so `fmt` below has a sibling with the same name.
impl std::fmt::Display for Config {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str("config")
    }
}

/// The sibling: same type, same method name, different trait.
impl std::fmt::Debug for Config {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str("config")
    }
}

/// A trait impl on a reference, which is not the same impl as the one on the
/// type.
impl Handler for &Key {
    fn handle(&self, key: &Key) -> bool {
        !self.is_empty() && !key.is_empty()
    }
}
